File: bindings/python/src/sparcli/humanize.py
"""Human-readable formatting: byte sizes, durations, relative time, numbers.

Each function returns a Python ``str``. Mirrors the C ``sc_humanize_*`` API.
"""

from __future__ import annotations

from dataclasses import dataclass

from ._ffi import ffi, lib
from .enums import ByteUnit


@dataclass
class HumanizeOpts:
    """Formatting options (defaults match the C zero-init behavior)."""

    decimals: int = 0
    decimal_sep: str | None = None   # None = '.'  (use ',' for de_DE)
    group_sep: str | None = None     # None = ','  (use '.' for de_DE)
    no_space: bool = False

    def _fill(self, c) -> None:
        c.decimals = self.decimals
        c.decimal_sep = self.decimal_sep.encode()[:1] if self.decimal_sep \
            else b"\x00"
        c.group_sep = self.group_sep.encode()[:1] if self.group_sep \
            else b"\x00"
        c.no_space = self.no_space


def _take(out) -> str:
    """Decode a C-heap string and free it."""
    if out == ffi.NULL:
        raise MemoryError("humanize: allocation failed")
    try:
        return ffi.string(out).decode()
    finally:
        lib.free(out)


def bytes(n: int, unit: ByteUnit = ByteUnit.SI,
          opts: HumanizeOpts | None = None) -> str:
    """Format a byte count, e.g. ``bytes(1536)`` -> ``"1.5 KB"``."""
    if opts is None:
        return _take(lib.sc_humanize_bytes(n, int(unit)))
    c = ffi.new("ScHumanizeOpts *")
    opts._fill(c)
    return _take(lib.sc_humanize_bytes_opts(n, int(unit), c[0]))


def number(value: float, opts: HumanizeOpts | None = None) -> str:
    """Format a number with thousands grouping, e.g. ``"1,234,567"``."""
    c = ffi.new("ScHumanizeOpts *")
    (opts or HumanizeOpts())._fill(c)
    return _take(lib.sc_humanize_number(value, c[0]))


def compact(value: float, opts: HumanizeOpts | None = None) -> str:
    """Format a number compactly, e.g. ``12400`` -> ``"12.4k"``."""
    c = ffi.new("ScHumanizeOpts *")
    (opts or HumanizeOpts())._fill(c)
    return _take(lib.sc_humanize_compact(value, c[0]))


def percent(ratio: float, opts: HumanizeOpts | None = None) -> str:
    """Format a ratio as a percentage, e.g. ``0.42`` -> ``"42%"``."""
    c = ffi.new("ScHumanizeOpts *")
    (opts or HumanizeOpts())._fill(c)
    return _take(lib.sc_humanize_percent(ratio, c[0]))


def duration(seconds: float) -> str:
    """Format a duration as two units, e.g. ``93`` -> ``"1m 33s"``."""
    return _take(lib.sc_humanize_duration(seconds))


def duration_clock(seconds: float) -> str:
    """Format a duration as a clock value, e.g. ``3725`` -> ``"01:02:05"``."""
    return _take(lib.sc_humanize_duration_clock(seconds))


def relative(when: int, now: int) -> str:
    """Format ``when`` relative to ``now`` (Unix seconds), e.g. "3 hours ago"."""
    return _take(lib.sc_humanize_relative(when, now))

File: bindings/rust/sparcli/src/humanize.rs
//! Human-readable formatting: byte sizes, durations, relative time and
//! grouped/compact numbers. Each function returns an owned `String`.

use crate::take_c_string;
use sparcli_sys as ffi;

repr_enum!(
    /// Unit system for [`bytes`]: `Si` (1000-based `KB`), `Iec` (1024-based
    /// `KiB`), `IecShort` (1024-based short `K`/`M`/`G`).
    ByteUnit {
        Si = ffi::ScByteUnit_SC_BYTES_SI,
        Iec = ffi::ScByteUnit_SC_BYTES_IEC,
        IecShort = ffi::ScByteUnit_SC_BYTES_IEC_SHORT,
    } default Si
);

/// Formatting options (zero-init = sensible defaults).
#[derive(Clone, Debug, Default)]
pub struct HumanizeOpts {
    /// Fixed number of fractional digits; `0` = per-function default.
    pub decimals: i32,
    /// Decimal separator; `None` = `'.'` (use `','` for de_DE).
    pub decimal_sep: Option<char>,
    /// Thousands separator for [`number`]; `None` = `','` (use `'.'` for de_DE).
    pub group_sep: Option<char>,
    /// Suppress the space between number and unit.
    pub no_space: bool,
}

impl HumanizeOpts {
    pub fn new() -> Self {
        HumanizeOpts::default()
    }
    pub fn decimals(mut self, n: i32) -> Self {
        self.decimals = n;
        self
    }
    pub fn decimal_sep(mut self, c: char) -> Self {
        self.decimal_sep = Some(c);
        self
    }
    pub fn group_sep(mut self, c: char) -> Self {
        self.group_sep = Some(c);
        self
    }
    pub fn no_space(mut self, on: bool) -> Self {
        self.no_space = on;
        self
    }
    fn raw(&self) -> ffi::ScHumanizeOpts {
        ffi::ScHumanizeOpts {
            decimals: self.decimals,
            decimal_sep: sep_byte(self.decimal_sep),
            group_sep: sep_byte(self.group_sep),
            no_space: self.no_space,
        }
    }
}

/// Maps an optional ASCII separator char to the C `char` field (`0` = unset).
fn sep_byte(c: Option<char>) -> std::os::raw::c_char {
    match c {
        Some(ch) => ch as u8 as std::os::raw::c_char,
        None => 0,
    }
}

/// Formats a byte count, e.g. `bytes(1536, ByteUnit::Si)` → `"1.5 KB"`.
pub fn bytes(n: u64, unit: ByteUnit) -> String {
    bytes_opts(n, unit, HumanizeOpts::default())
}

/// Options-taking variant of [`bytes`].
pub fn bytes_opts(n: u64, unit: ByteUnit, opts: HumanizeOpts) -> String {
    unsafe { take_c_string(ffi::sc_humanize_bytes_opts(n, unit.raw(), opts.raw())) }
}

/// Formats a number with thousands grouping, e.g. `1234567` → `"1,234,567"`.
pub fn number(value: f64, opts: HumanizeOpts) -> String {
    unsafe { take_c_string(ffi::sc_humanize_number(value, opts.raw())) }
}

/// Formats a number compactly, e.g. `12400` → `"12.4k"`.
pub fn compact(value: f64, opts: HumanizeOpts) -> String {
    unsafe { take_c_string(ffi::sc_humanize_compact(value, opts.raw())) }
}

/// Formats a ratio as a percentage, e.g. `0.42` → `"42%"`.
pub fn percent(ratio: f64, opts: HumanizeOpts) -> String {
    unsafe { take_c_string(ffi::sc_humanize_percent(ratio, opts.raw())) }
}

/// Formats a duration as two units, e.g. `93` → `"1m 33s"`.
pub fn duration(seconds: f64) -> String {
    unsafe { take_c_string(ffi::sc_humanize_duration(seconds)) }
}

/// Formats a duration as a clock value, e.g. `3725` → `"01:02:05"`.
pub fn duration_clock(seconds: f64) -> String {
    unsafe { take_c_string(ffi::sc_humanize_duration_clock(seconds)) }
}

/// Formats `when` relative to `now` (Unix seconds), e.g. `"3 hours ago"`.
pub fn relative(when: i64, now: i64) -> String {
    unsafe {
        take_c_string(ffi::sc_humanize_relative(
            when as ffi::time_t,
            now as ffi::time_t,
        ))
    }
}
